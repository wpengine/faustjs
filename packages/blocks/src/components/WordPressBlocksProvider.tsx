import React, { FC } from 'react';
import type { DocumentNode } from '@apollo/client';
import { BlocksTheme } from '../types/theme.js';

export type WordPressBlockBase = React.FC & {
  fragments: {
    key: string;
    entry: DocumentNode;
  };
  displayName: string;
  name: string;
  config: {
    name: string;
  };
};

/**
 * WordPressBlock is a React component that contains some optional properties that we are
 * used to match it with equivalent block data from the API
 */
export type WordPressBlock<P = { [key: string]: any }> = FC<P> &
  Partial<
    Pick<WordPressBlockBase, 'config' | 'displayName' | 'name' | 'fragments'>
  >;

export type WordPressBlocksContextType =
  | { [key: string]: WordPressBlock }
  | undefined;
export type WordPressThemeContextType = BlocksTheme | undefined;

export const WordPressBlocksContext =
  React.createContext<WordPressBlocksContextType>(undefined);
export const WordPressThemeContext =
  React.createContext<WordPressThemeContextType>(undefined);

export type WordPressBlocksProviderConfig = {
  blocks: { [key: string]: WordPressBlock };
  theme?: BlocksTheme;
};

/**
 * WordPressBlocksProvider is used as a central store for the available list of WordPressBlock types.
 * @param props
 * @returns
 */
export function WordPressBlocksProvider(props: {
  children: React.ReactNode;
  config: WordPressBlocksProviderConfig;
}) {
  const { children, config } = props;
  const { blocks, theme } = config;

  return (
    <WordPressBlocksContext.Provider value={blocks}>
      <WordPressThemeContext.Provider value={theme}>
        {children}
      </WordPressThemeContext.Provider>
    </WordPressBlocksContext.Provider>
  );
}

/**
 * useBlocksTheme can be used to retrieve the theme
 * from within the WordPressBlocksProvider.
 *
 * @example
 * ```
 * const theme = useBlocksTheme();
 * ```
 */
export function useBlocksTheme(): BlocksTheme {
  // If it's an empty object, the provider hasn't been initialized.Â±
  if (
    typeof WordPressBlocksContext === 'undefined' ||
    typeof WordPressBlocksContext === 'undefined'
  ) {
    throw new Error(
      'useBlocksTheme hook was called outside of context, make sure your app is wrapped with WordPressBlocksProvider',
    );
  }

  const themeContext = React.useContext(WordPressThemeContext);
  if (typeof themeContext === 'undefined') {
    throw new Error(
      'useBlocksTheme hook was called outside of context, make sure your app is wrapped with WordPressBlocksProvider',
    );
  }

  return themeContext;
}
