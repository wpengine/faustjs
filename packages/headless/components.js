module.exports = require('./dist/components');
