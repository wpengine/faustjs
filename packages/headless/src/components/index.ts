import Menu from './menu/Menu';
import MenuItem from './menu/MenuItemInterface';
import TemplateLoader from './TemplateLoader';
import NextTemplateLoader from './NextTemplateLoader';
import WPHead from './WPHead';

export { Menu, MenuItem, TemplateLoader, WPHead, NextTemplateLoader };
