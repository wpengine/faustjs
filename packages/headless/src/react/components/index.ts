export * from './menu';
export * from './TemplateLoader';
