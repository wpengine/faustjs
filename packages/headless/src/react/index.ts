export * from './hooks';
export * from './provider';
export * from './components';
