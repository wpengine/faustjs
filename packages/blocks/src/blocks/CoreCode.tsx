import { gql } from '@apollo/client';
import React from 'react';
import { useBlocksTheme } from '../components/WordPressBlocksProvider.js';
import {
  BlockWithAttributes,
  ContentBlock,
} from '../components/WordPressBlocksViewer.js';
import { getStyles } from '../utils/index.js';

export type CoreCodeFragmentProps = ContentBlock & {
  attributes: {
    anchor?: string;
    backgroundColor?: string;
    borderColor?: string;
    className?: string;
    content?: string;
    cssClassName?: string;
    fontFamily?: string;
    fontSize?: string;
    gradient?: string;
    lock?: string;
    style?: string;
    textColor?: string;
  };
};

export function CoreCode(props: BlockWithAttributes<CoreCodeFragmentProps>) {
  const theme = useBlocksTheme();
  const style = getStyles(theme, { ...props });
  const { attributes } = props;

  return (
    <pre style={style} className={attributes?.cssClassName}>
      <code
        // eslint-disable-next-line react/no-danger
        dangerouslySetInnerHTML={{ __html: attributes?.content ?? '' }}
      />
    </pre>
  );
}

CoreCode.fragments = {
  key: `CoreCodeBlockFragment`,
  entry: gql`
    fragment CoreCodeBlockFragment on CoreCode {
      attributes {
        anchor
        backgroundColor
        borderColor
        className
        content
        cssClassName
        fontFamily
        fontSize
        gradient
        lock
        style
        textColor
      }
    }
  `,
};
