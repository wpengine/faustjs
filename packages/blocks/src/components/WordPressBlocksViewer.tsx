import React from 'react';
import resolveBlockTemplate from '../resolveBlockTemplate.js';
import { WordPressBlocksContext } from './WordPressBlocksProvider.js';

export interface WordpressBlocksViewerProps {
  /**
   * Block data obtained from WPGraphQL Content Blocks
   */
  blocks: Array<ContentBlock | null>;
  /**
   * An optional React component that will be used if
   * no React components are resolved for a given block.
   *
   * @example
   * ```jsx
   * const MyFallbackComponent = ({renderedHtml}) => {
   *   return (
   *     <span dangerouslySetInnerHTML={{__html: renderedHtml}} />
   *   )
   * }
   *
   * <WordPressBlocksViewer fallbackBlock={MyFallbackComponent} />
   * ```
   */
  fallbackBlock?: React.FC<ContentBlock>;
}

export interface ContentBlock {
  __typename?: string;
  apiVersion?: number;
  cssClassNames?: string;
  innerBlocks?: Array<ContentBlock | null>;
  isDynamic?: boolean;
  name?: string;
  renderedHtml?: string;
}

export interface BlockWithAttributes extends ContentBlock {
  attributes?: Record<string, unknown> | null;
}

/**
 * WordPressBlocksViewer is the main component that renders blocks taken from WordPress as a list of ContentBlock[] data.
 * @param props WordpressBlocksViewerProps
 * @returns JSX.Component that renders the block tree.
 */
export function WordPressBlocksViewer(props: WordpressBlocksViewerProps) {
  const blocks = React.useContext(WordPressBlocksContext);

  if (!blocks) {
    throw new Error('Blocks are required. Please add them to your config.');
  }

  const { blocks: editorBlocks, fallbackBlock } = props;

  if (!editorBlocks) {
    throw new Error('The "blocks" prop is required in <WordPressBlocksViewer>');
  }

  const renderedBlocks = editorBlocks.map((blockProps, idx) => {
    const BlockTemplate = resolveBlockTemplate(
      blockProps,
      blocks,
      fallbackBlock,
    );

    return React.createElement<ContentBlock>(BlockTemplate, {
      ...blockProps,
      // eslint-disable-next-line react/no-array-index-key
      key: idx,
    });
  });
  // eslint-disable-next-line react/jsx-no-useless-fragment
  return <>{renderedBlocks}</>;
}
