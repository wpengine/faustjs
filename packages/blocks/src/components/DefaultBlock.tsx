import React from 'react';
import { ContentBlock } from './WordPressBlocksViewer.js';

/**
 * DefaultBlock is an instance of WordPressBlock that is used in case no matching component is found when rendering the list of blocks.
 * @param param0
 * @returns JSX.Element
 */
export default function DefaultBlock({ renderedHtml }: ContentBlock) {
  // eslint-disable-next-line react/no-danger
  return <div dangerouslySetInnerHTML={{ __html: renderedHtml ?? '' }} />;
}

DefaultBlock.displayName = 'DefaultBlock';
