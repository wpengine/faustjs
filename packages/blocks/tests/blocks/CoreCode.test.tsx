/** @jest-environment jsdom */
import * as React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { WordPressBlocksProvider } from '../../src/components/WordPressBlocksProvider';
import { CoreCode, CoreCodeFragmentProps } from '../../src/blocks/CoreCode.js';
import { BlockWithAttributes } from '../../src/components/WordPressBlocksViewer';

function renderProvider(props: BlockWithAttributes<CoreCodeFragmentProps>) {
  return render(
    <WordPressBlocksProvider config={{ blocks: [], theme: {} }}>
      <CoreCode {...props} />
    </WordPressBlocksProvider>,
  );
}

describe('<CoreCode />', () => {
  test('renders the component correctly', () => {
    renderProvider({
      attributes: { content: '&lt;div&gt;My code block here&lt;/div&gt;' },
    });
    expect(
      screen.queryByText('<div>My code block here</div>'),
    ).toBeInTheDocument();
  });

  test('applies the correct styles', () => {
    const tree = renderProvider({
      attributes: {
        backgroundColor: 'vivid-red',
        borderColor: 'pale-cyan-blue',
        content: '&lt;div&gt;My code block&lt;/div&gt;',
        cssClassName:
          'wp-block-code has-border-color has-pale-cyan-blue-border-color has-background-color has-vivid-red-background-color has-text-color has-background has-medium-font-size',
        fontSize: 'medium',
        style: '{"border":{"radius":"37px","width":"27px"}}',
        textColor: 'background',
      },
    });

    expect(tree.container).toMatchSnapshot(`
    <div>
      <pre
        class="wp-block-code has-border-color has-pale-cyan-blue-border-color has-background-color has-vivid-red-background-color has-text-color has-background has-medium-font-size"
        style="border-width: 27px; border-radius: 37px;"
      >
        <code>
          &lt;div&gt;My code block&lt;/div&gt;
        </code>
      </pre>
    </div>    
    `);
  });
});
