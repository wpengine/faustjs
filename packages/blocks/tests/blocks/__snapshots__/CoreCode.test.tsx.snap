// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoreCode /> applies the correct styles: 
    <div>
      <pre
        class="wp-block-code has-border-color has-pale-cyan-blue-border-color has-background-color has-vivid-red-background-color has-text-color has-background has-medium-font-size"
        style="border-width: 27px; border-radius: 37px;"
      >
        <code>
          &lt;div&gt;My code block&lt;/div&gt;
        </code>
      </pre>
    </div>    
     1`] = `
<div>
  <pre
    class="wp-block-code has-border-color has-pale-cyan-blue-border-color has-background-color has-vivid-red-background-color has-text-color has-background has-medium-font-size"
    style="border-width: 27px; border-radius: 37px;"
  >
    <code>
      &lt;div&gt;My code block&lt;/div&gt;
    </code>
  </pre>
</div>
`;
